python_modules/dagster/dagster_tests/execution_tests/dynamic_tests/test_reexecute.py

```diff
@@ -139,3 +147,157 @@ def test_reexec_from_parent_3():
+
+
+def dynamic_with_transitive_optional_output_job():
+    @op(out=Out(is_required=False))
+    def add_one_with_optional_output(context, i: int):
+        if (
+            context.pipeline_run.parent_run_id
+            and i % 2 == 0  # re-execution run skipped odd numbers
+            or not context.pipeline_run.parent_run_id
+            and i % 2 == 1  # root run skipped even numbers
+        ):
+            yield Output(i + 1)
+
+    @job(resource_defs={"io_manager": fs_io_manager})
+    def _dynamic_with_transitive_optional_output_job():
+        dynamic_results = dynamic_op().map(lambda n: echo(add_one_with_optional_output(n)))
+        adder(dynamic_results.collect())
+
+    return _dynamic_with_transitive_optional_output_job
+
+
+def test_reexec_dynamic_with_transitive_optional_output_job_1():
+    with instance_for_test() as instance:
+        result = dynamic_with_transitive_optional_output_job().execute_in_process(instance=instance)
+        assert result.success
+        assert result.output_for_node("adder") == sum([i + 1 for i in range(10) if i % 2 == 1])
+
+        # re-execute all
+        re_result = reexecute_pipeline(
+            reconstructable(dynamic_with_transitive_optional_output_job),
+            parent_run_id=result.run_id,
+            instance=instance,
+        )
+        assert re_result.success
+        assert re_result.output_for_solid("adder") == sum([i + 1 for i in range(10) if i % 2 == 0])
+
+
+def test_reexec_dynamic_with_transitive_optional_output_job_2():
+    with instance_for_test() as instance:
+        result = dynamic_with_transitive_optional_output_job().execute_in_process(instance=instance)
+
+        # re-execute the step where the source yielded an output
+        re_result = reexecute_pipeline(
+            reconstructable(dynamic_with_transitive_optional_output_job),
+            parent_run_id=result.run_id,
+            instance=instance,
+            step_selection=["echo[1]"],
+        )
+        assert re_result.success
+        assert re_result.result_for_solid("echo").output_value() == {"1": 2}
+
+
+def test_reexec_dynamic_with_transitive_optional_output_job_3():
+    with instance_for_test() as instance:
+        result = dynamic_with_transitive_optional_output_job().execute_in_process(instance=instance)
+
+        # re-execute the step where the source did not yield
+        re_result = reexecute_pipeline(
+            reconstructable(dynamic_with_transitive_optional_output_job),
+            parent_run_id=result.run_id,
+            instance=instance,
+            step_selection=["echo[0]"],
+            raise_on_error=False,
+        )
+        # when all the previous runs have skipped yielding the source,
+        # run would fail because of run_id returns None
+        # FIXME: https://github.com/dagster-io/dagster/issues/3511
+        # ideally it should skip the step because all its previous runs have skipped and finish the run successfully
+        assert not re_result.success
```

Comment:
would like to get opinions on what the right behavior should be.

![image](https://user-images.githubusercontent.com/4531914/142153510-8638a053-98df-4d6d-bc8d-d2c635c8afc2.png)
this is re-executing the step `echo[0]` where the source doesn't yield any outputs (i.e. `echo[0]` was skipped in the parent run). as a result, in [1], the source run id is None which will cause this error:
![image](https://user-images.githubusercontent.com/4531914/142154361-1e269c7c-0bf6-4dce-b0a4-8d98a8768ef5.png)

I think fixing would involve deeper changes, i.e. https://github.com/dagster-io/dagster/issues/3511, and I'd like to unblock the user reported issue asap. So, my proposal is to add a warning in [1] in the short term because we can't simply error or skip based on no output events found (which is also true when someone backfill a selection with fixed path io manager).

But I think it's reached a point where the re-execution source run id logic is complicated enough that we shouldn't carry this tech debt further. So I plan to start https://github.com/dagster-io/dagster/issues/3511 next week.

Reply:
In terms of the desired behavior: if "echo[0]" never executed, we shouldn't allow re-execution of it, right?

Reply:
I think when `echo[0]` is the only step, it makes sense to disallow re-execution. but in cases like this:
<img width="965" alt="Screen Shot 2021-11-17 at 9 59 02 AM" src="https://user-images.githubusercontent.com/4531914/142256376-9ffaac79-1f4c-4efa-8e48-523b4adbcad9.png">
 i think the desired behavior for re-executing `echo[0]*` should be allow the re-execution and skip the echo[0]

Reply:
Do you mean allow re-executing `echo*` (not `echo[0]*`)?

Reply:
i meant `echo[0]*` - i'm imagining a case where echo[1]...echo[9] all succeeded but echo[0] skipped (in non deterministic way like this bug report), and then the user could want to re-execute *from* echo[0] to finish the whole job.

my hypothesis is it's rare but it could still happen, so it may not be worth adding additional logic in [1] to handle this edge case, but it's still nice to throw some warnings so there's a clue what could be wrong when the run fails.

Reply:
> if echo[0] skips, it's because the step upstream of echo[0] skipped, right? 
yes
> so, for echo[0] not to skip, the re-execution, would need to start with the upstream step, right?
the part im not sure is whether we want echo[0] not to skip - imo the re-execution would skip echo[0] and proceed with the downstream. in other words, failing may not be dieal.

Reply:
IMO it just doesn't make any sense to "re-execute" a step that didn't execute in the first place.

Reply:
ok - tbh i dont have strong opinions and i think in order to error it properly we need to a some refactor, which doesn't necessarily need to block this fix. lets discuss this use case more later.

the question here is whether the warning would make sense or we should just do nothing and fix this behavior later.

Reply:
Gotcha - yeah, I didn't mean to imply that we need to fix that as part of this PR.  IMO a warning is better than nothing.